### ISSUE TITLE
RAC 3: Preferred Way To Handle Enabled/Disable of UIControl?

### ISSUE THREAD

**Issue by aventurella** [ReactiveCocoa/ReactiveCocoa#2250]
I have 2 options that work, and I don't know if there are any others, so I thought I would present them and ask.

- Option 1: DynamicProperty
- Option 2: Action.enableIf

### Option 1 DynamicProperty:

```swift
import Foundation
import UIKit
import ReactiveCocoa

struct MyViewModel{
    let label = MutableProperty("")
}

class MyController: UIViewController{

    @IBOutlet weak var labelField: UITextField!
    @IBOutlet weak var saveButton: UIControl!

    var viewModel: MyViewModel!
    var saveClicked: CocoaAction!

     /*>>>> HERE <<<<*/
    private lazy var saveEnabled: DynamicProperty = {
        return DynamicProperty(object: self.saveButton, keyPath: "enabled")
    }()

    override func viewDidLoad() {
        super.viewDidLoad()

        viewModel = MyViewModel()

        // listen to the labelField
        let value = labelField.rac_textSignal().toSignalProducer()
                .map{ $0 as! String }
                .flatMapError{ err in SignalProducer<String, NoError>(value: "")}

        // nothing action
        let action = Action<Void, Void, NoError>{
            SignalProducer.empty
        }

        saveClicked = CocoaAction(action){ _ in
                print("HEY-O!")
        }

        saveButton.addTarget(saveClicked, action: CocoaAction.selector, forControlEvents: UIControlEvents.TouchUpInside)

        // enable/disable the button
        saveEnabled <~ value.map{ $0.unicodeScalars.count > 2 }

        // update the viewModel
        viewModel.label <~ value
    }
}
```

### Option 2 Action.enableIf:

```swift
struct MyViewModel{
    let label = MutableProperty("")
}

class MyController: UIViewController{

    @IBOutlet weak var labelField: UITextField!
    @IBOutlet weak var saveButton: UIControl!

    var viewModel: MyViewModel!
    var saveClicked: CocoaAction!
    var saveEnabled = MutableProperty(false)

    override func viewDidLoad() {
        super.viewDidLoad()

        viewModel = MyViewModel()

        let value = labelField.rac_textSignal().toSignalProducer()
                .map{ $0 as! String }
                .flatMapError{ err in SignalProducer<String, NoError>(value: "")}
         
       /*>>>> HERE <<<<*/
        let action = Action<Void, Void, NoError>(enabledIf: saveEnabled){
            SignalProducer.empty
        }

        saveClicked = CocoaAction(action){ _ in
            /*>>>> AND  HERE <<<<*/
            if action.enabled.value{
                print("HEY-O!")
            }
        }

        saveButton.addTarget(saveClicked, action: CocoaAction.selector, forControlEvents: UIControlEvents.TouchUpInside)

        // enable/disable the button
        saveEnabled <~ value.map{ $0.unicodeScalars.count > 2 }

        // update the viewModel
        viewModel.label <~ value
    }
}
```

**Comment by aventurella** [comment]
Digging in some more, if a `CocoaAction` wraps an `Action` my assumption is that wrapper (CocoaAction) is there provide convince to attach to `UIControl` subclasses on iOS.  Looking at the source that seems reasonable:

```swift

public init<Input, Output, Error>(_ action: Action<Input, Output, Error>, _ inputTransform: AnyObject? -> Input) {
	_execute = { input in
		let producer = action.apply(inputTransform(input))
		producer.start()
	}
    ...
}

There are conveniences like `unsafeCocoaAction` on `Action` that strengthen this idea for me. However, there is no clear place that dictates where I should be actually writing my response to the action.

Using `unsafeCocoaAction` is out unless you reset the property to be something of your own design. Additionally, CocoaAction does not expose the producer, so you can't attach `next:` there.  

`Action` is supposed to return a `SignalProducer<Output, Error>` so it seems silly add the actual code for the action there. You can't attach to the `next:` of an action either without `apply`ing it first, which just executes immediately.

I concede I am likely missing something obvious, so I hope I can get filled in. From looking at it tough, it does not seems like there is any designated space to actually add code to represent the action.  Can anyone clear this up for me?

**Comment by aventurella** [comment]
Ok.. I think I got it... I missed `Action.values`


So if you wanted to simulate what an `IBAction` does with Action/Cocoa Action it would look like this:

```swift
@IBOutlet weak var myButton: UIControl!

func setup(){
    let action = Action<AnyObject, AnyObject, NoError>{ button in
      SignalProducer<AnyObject, NoError>(value: button)
    }

    myButton.addTarget(action.unsafeCocoaAction, action: CocoaAction.selector, forControlEvents: UIControlEvents.TouchUpInside)

    action.values.observe(next: buttonClicked)
}

func buttonClicked(sender: AnyObject){
    print(sender)
}

```

**Comment by startupthekid** [comment]
I do it a little bit differently:

```
// Action for a button that does nothing
private lazy var confirmEnabled: AnyProperty<Bool> = {
        return AnyProperty(initialValue: false, producer: self.buttonEnabled.producer)
    }()

private lazy var confirmAction: Action<Void, Void, NoError> = {
        return Action<Void, Void, NoError>(enabledIf: self.confirmEnabled) { _ in return SignalProducer.empty }
    }()
 
// CocoaAction wrapper
private lazy var confirmCocoaAction: CocoaAction = {
    return CocoaAction(self.confirmAction, input: ())
 }()

// Add the action
self.button.addTarget(self.confirmCocoaAction, action: CocoaAction.selector, forControlEvents: .TouchUpInside)
```

Now so far, we have a property that enables and disables the action, the action, and the cocoa wrapper. But keep in mind, the `Action` doesn't actually effect the `enabled` property on the button.
So to achieve that just add one more line:

```
self.confirmAction.enabled.producer.startWithNext{ self.button.enabled = $0 }
```
It's important to remember to use the producer of the enabled property, not the signal because the signal won't send past values.

**Comment by gbuela** [comment]
In case you guys are interested, I came up with a wrapper over CocoaAction and UIControl in [this project](https://github.com/gbuela/github-users).

I use it like this:

```swift
        self.fetchControlAction = CocoaControlAction(control: self.goButton,
                                                     action: self.viewModel.fetchUsersAction,
                                                     input: "")
```

This automatically connects the .TouchUpInside event (or a different one if you specify it) of your control to the Action, and by default synchronizes the control's enabled state to match the Action's.

Here I gave the ViewModel the responsibility of defining the Action and the criteria by which it should be enabled.

Then I use this convenience method to observe completion. 
```swift
        self.fetchControlAction
            .observeActionComplete { [weak self] _ in
                self?.usersTable.reloadData()
        }
```
This does a few things for me: observeNext, filter by .Complete event, observe on UIScheduler. 

I think my view controller ended up a lot cleaner after moving somewhere else all the boilerplate for this common scenario.
